File: core/src/main/java/com/gitlab/aecsocket/sokol/core/stat/Stat.java
package com.gitlab.aecsocket.sokol.core.stat;

import com.gitlab.aecsocket.minecommons.core.Validation;
import com.gitlab.aecsocket.minecommons.core.translation.Localizer;
import com.gitlab.aecsocket.sokol.core.Renderable;
import net.kyori.adventure.text.Component;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.jetbrains.annotations.NotNull;
import org.spongepowered.configurate.ConfigurationNode;
import org.spongepowered.configurate.serialize.SerializationException;

import java.lang.reflect.Type;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

public interface Stat<T> extends Renderable {
    String key();
    Optional<T> defaultValue();

    Value<T> deserialize(Type type, ConfigurationNode node) throws SerializationException;

    Node<T> node(Value<T> value);
    Component renderValue(Locale locale, Localizer lc, T value);

    static String renderKey(String key) { return "stat." + key + ".name"; }
    static String renderFormatKey(String key) { return "stat." + key + ".format"; }

    @Override
    default Component render(Locale locale, Localizer lc) {
        return lc.safe(locale, renderKey(key()));
    }

    default Component renderFormat(Locale locale, Localizer lc, Component value) {
        return lc.safe(locale, renderFormatKey(key()),
                "value", value);
    }

    interface Value<T> extends Renderable {
        T compute(T cur);
        boolean discardsPrevious();
    }

    interface InitialValue<T> extends Value<T> {
        T first();
    }

    final class Node<T> implements Iterable<Node<T>> {
        private final Stat<T> stat;
        private final Value<T> value;
        private @Nullable Node<T> next;

        public Node(Stat<T> stat, Value<T> value) {
            if (!(value instanceof InitialValue))
                throw new IllegalStateException("Attempting to start stat node chain with non-initial value");
            this.stat = stat;
            this.value = value;
        }

        public Node(Node<T> o) {
            stat = o.stat;
            value = o.value;
            next = o.next == null ? null : new Node<>(o.next);
        }

        public Stat<T> stat() { return stat; }
        public Value<T> value() { return value; }

        public T compute() {
            if (!(value instanceof InitialValue<T> initial))
                throw new IllegalStateException("Start of stat node chain is a non-initial value");
            Node<T> node = next;
            T value = initial.first();
            for (; node != null; node = node.next)
                value = node.value.compute(value);
            return value;
        }

        public @Nullable Node<T> next() { return next; }
        public Node<T> chain(Node<T> next) {
            Validation.assertNot(next == this, "Next node in chain is itself");
            if (this.next == null)
                this.next = next;
            else
                this.next.chain(next);
            return this;
        }

        public List<Node<T>> asList() {
            List<Node<T>> result = new LinkedList<>();
            Node<T> node = this;
            for (; node != null; node = node.next)
                result.add(node);
            return result;
        }

        @NotNull
        @Override
        public Iterator<Node<T>> iterator() {
            return new Iterator<Node<T>>() {
                private Node<T> cur = Node.this;

                @Override
                public boolean hasNext() {
                    return cur != null;
                }

                @Override
                public Node<T> next() {
                    Node<T> ret = cur;
                    cur = cur.next;
                    return ret;
                }
            };
        }

        public Stream<Node<T>> stream() {
            return StreamSupport.stream(spliterator(), false);
        }

        @Override
        public String toString() {
            return "[" + value + "]" + (next == null ? "" : " -> " + next);
        }
    }

    final class OperationDeserializer<T> {
        @FunctionalInterface
        public interface Operation<T> {
            Value<T> createValue(Type type, ConfigurationNode node, ConfigurationNode... args) throws SerializationException;
        }

        private record OperationData<T>(
                Operation<T> operation,
                String[] fields
        ) {}

        private final Map<String, OperationData<T>> operations;
        private final @Nullable String defaultOperation;

        private OperationDeserializer(Map<String, OperationData<T>> operations, @Nullable String defaultOperation) {
            this.operations = Collections.unmodifiableMap(operations);
            this.defaultOperation = defaultOperation;
        }

        private String formatArgs(Class<?>... types) {
            return Stream.of(types).map(Class::getSimpleName).collect(Collectors.joining(", "));
        }

        public Value<T> deserialize(Type type, ConfigurationNode node) throws SerializationException {
            List<? extends ConfigurationNode> nodes = node.isList() ? new ArrayList<>(node.childrenList()) : null;
            OperationData<T> op;
            if (nodes != null && nodes.size() > 0 && nodes.get(0).raw() instanceof String opName) {
                op = operations.get(opName);
                if (op == null)
                    throw new SerializationException(node, type, "Invalid operation type '" + opName + "'");
                nodes.remove(0);

                if (op.fields.length != nodes.size())
                    throw new SerializationException(node, type, "Operation '" + opName + "' requires fields: [" +
                            String.join(", ", op.fields) + "], passed " + nodes.size());
                return op.operation.createValue(type, node, nodes.toArray(new ConfigurationNode[0]));
            } else {
                op = operations.get(defaultOperation);
                if (op == null)
                    throw new SerializationException(node, type, "No operation provided");
                if (op.fields.length != 1)
                    throw new SerializationException(node, type, "Operation '" + defaultOperation + "' requires fields: [" +
                            String.join(", ", op.fields) + "]");
                return op.operation.createValue(type, node, node);
            }
        }

        public static final class Builder<T> {
            private final Map<String, OperationData<T>> operations = new HashMap<>();
            private @Nullable String defaultOperation;

            public Builder<T> operation(String key, Operation<T> operation, String... fields) {
                operations.put(key, new OperationData<>(operation, fields));
                return this;
            }

            public Builder<T> defaultOperation(String defaultOperation) {
                this.defaultOperation = defaultOperation;
                return this;
            }

            public OperationDeserializer<T> build() {
                return new OperationDeserializer<>(operations, defaultOperation);
            }
        }

        public static <T> Builder<T> builder() {
            return new Builder<>();
        }
    }
}

File: core/src/main/kotlin/com/gitlab/aecsocket/sokol/core/stat/Stat.kt
package com.gitlab.aecsocket.sokol.core.stat

import com.gitlab.aecsocket.alexandria.core.extension.force
import com.gitlab.aecsocket.alexandria.core.extension.forceList
import net.kyori.adventure.key.Key
import org.spongepowered.configurate.ConfigurationNode
import org.spongepowered.configurate.serialize.SerializationException

interface Stat<T : Any> {
    val key: Key

    fun deserialize(node: ConfigurationNode): Value<T>

    fun interface Value<T : Any> {
        fun next(last: T): T

        interface First<T : Any> : Value<T> {
            fun first(): T
        }

        interface Discarding<T : Any> : First<T> {
            override fun next(last: T) = first()
        }
    }

    open class Node<T : Any>(
        val stat: Stat<T>,
        open val value: Value<T>,
        var next: Node<T>? = null
    ) {
        fun value(last: T): T {
            val res = value.next(last)
            return next?.value(res) ?: res
        }

        fun last(): Node<T> = next?.last() ?: this

        fun add(node: Node<T>) {
            last().next = node
        }

        override fun toString() = next?.let { "$value -> ${it.value}" } ?: value.toString()

        class First<T : Any>(
            stat: Stat<T>,
            override val value: Value.First<T>,
            next: Node<T>? = null
        ) : Node<T>(stat, value, next) {
            fun compute(): T {
                val res = value.first()
                return next?.value(res) ?: res
            }

            override fun toString() = "First(${super.toString()})"
        }
    }
}

abstract class AbstractStat<T : Any>(
    namespace: String, key: String
) : Stat<T> {
    override val key = Key.key(namespace, key)
}


private fun <T : Any> statNodeOfInternal(stat: Stat<T>, values: Collection<Stat.Value<T>>): Stat.Node<T>? {
    if (values.isEmpty()) return null
    val first = values.first()
    return if (first is Stat.Value.First) Stat.Node.First(stat, first, statNodeOfInternal(stat, values.drop(1)))
    else Stat.Node(stat, first, statNodeOfInternal(stat, values.drop(1)))
}

fun <T : Any> statNodeOf(stat: Stat<T>, values: Collection<Stat.Value<T>>): Stat.Node<T> {
    return statNodeOfInternal(stat, values) ?: throw IllegalArgumentException("No stat values passed")
}

fun statTypes(vararg stats: Stat<*>): Map<Key, Stat<*>> = stats.map {
    it.key to it
}.associate { it }

data class Operation<T : Any>(
    val argNames: List<String>,
    val action: (List<ConfigurationNode>) -> Stat.Value<T>
)

inline fun <reified T : Any> opDeserialize(
    node: ConfigurationNode,
    vararg operations: Pair<String, (List<ConfigurationNode>) -> Stat.Value<T>>,
): Stat.Value<T> {
    val type = T::class.java
    val list = node.forceList(type)
    if (list.isEmpty())
        throw SerializationException(node, type, "List must contain at least the operator")
    val opKey = list[0].force<String>()
    val opMap = operations.associate { it }
    val operation = opMap[opKey]
        ?: throw SerializationException(node, type, "Invalid operator '$opKey' for stat type $type")
    val args = list.drop(1)
    return operation(args)
}
